### PR TITLE
README Revisions

### PR DESCRIPTION
**Pull request by jackrueth**
This PR updates the main readme of the tutorials repository. 
* Updates the headers for each different way to access the Jupyter Notebooks
* Updated hyperlink to BeakerX
* Updated coloring on the tutorials map

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,37 +6,39 @@ This project contains example code for working with
...
+4. `conda env create -f environment.yml` to create a conda environment with the
+   dependencies these notebooks need.
+5. `conda activate scijava` to activate the environment.
+6. `jupyter notebook` to launch Jupyter Notebook in a web browser window.
+7. In the browser, click into `notebooks`, then click on the
+   `ImageJ-Tutorials-and-Demo.ipynb` notebook to open it.
 
-[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/imagej/tutorials/master)
-
+### (Easiest) Using this Repository
```
Comment (gselzer) [inline README.md:21]: I'm not sure I'd agree that this is the easiest way to view them. I'd actually suggest that the "Using Binder" option below is the easiest. 

What do you think is the best ordering of these options? I'd propose:
1. Using Binder
2. Using Miniconda
3. Viewing the static pages on nbviewer
4. Viewing the static pages on Github

File: README.md [hunk 2]
```diff
@@ -6,37 +6,39 @@ This project contains example code for working with
...
 The easiest way to get started with the ImageJ and SciJava APIs is via the
 [ImageJ Jupyter notebooks](https://imagej.github.io/tutorials),
 located in the `notebooks` subfolder of this repository.
 
+### Using Binder
+[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/imagej/tutorials/master)
+
 Use the "launch binder" badge above to try the Jupyter notebooks on the cloud
 using [Binder](https://mybinder.org), with no local installation necessary.
```
Comment (gselzer) [inline README.md:30]: I would mention that startup can take a minute.

File: README.md [hunk 3]
```diff
@@ -6,37 +6,39 @@ This project contains example code for working with
...
+5. `conda activate scijava` to activate the environment.
+6. `jupyter notebook` to launch Jupyter Notebook in a web browser window.
+7. In the browser, click into `notebooks`, then click on the
+   `ImageJ-Tutorials-and-Demo.ipynb` notebook to open it.
 
-[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/imagej/tutorials/master)
-
+### (Easiest) Using this Repository
 The easiest way to get started with the ImageJ and SciJava APIs is via the
 [ImageJ Jupyter notebooks](https://imagej.github.io/tutorials),
```
Comment (gselzer) [inline README.md:23]: Can we add instructions on how to get from the linked page to the notebook pages?

File: README.md [hunk 4]
```diff
@@ -6,37 +6,39 @@ This project contains example code for working with
...
 located in the `notebooks` subfolder of this repository.
 
+### Using Binder
+[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/imagej/tutorials/master)
+
 Use the "launch binder" badge above to try the Jupyter notebooks on the cloud
 using [Binder](https://mybinder.org), with no local installation necessary.
 
+### Using BeakerX
```
Comment (gselzer) [inline README.md:32]: Can we find a way to separate these informational sections from the viewing instructions? Another header would probably suffice.

**Review by gselzer — CHANGES_REQUESTED** [review]
This is a step in the right direction @jackrueth! The new graph looks great on all the Github themes!

File: README.md (hunk 1 above)
Comment (ctrueden) [reply README.md:21]: The options are complex and varied enough that perhaps we should have an introductory bullet list, _before_ breaking out into sections. Here is some proposed text:

<blockquote>

There are several ways to work with Jupyter notebooks, each with strengths and weaknesses:

| Method | Pros | Cons | Running live? |
|--------|------|------|---------------|
| Binder | Run code on the cloud, no local installation. | Takes a minute or three to spin up | Yes |
| Local environment | Run code locally on your machine. | Must install developer tools. | Yes |
| On GitHub | View notebooks quickly on GitHub's website. | Code does not run live; rendering is less complete & correct than on nbviewer. | No |
| nbviewer | View notebooks nicely rendered on nbviewer.org. | Code does not run live; not integrated with GitHub. | No |

See the following sections for details on each method.

</blockquote>

File: README.md (hunk 3 above)
Comment (ctrueden) [reply README.md:23]: The link `imagej.github.io/tutorials` needs to be changed to `imagej.github.io/tutorials/notebooks`

**Review by ctrueden — CHANGES_REQUESTED** [review]
Thanks for working on this, @jackrueth! I added some additional comments beneath @gselzer's comments. 

One additional thing: with ca075cfd952b3d02d632a95d35fbfa5dd8e3127e I don't see any significant different on the lines based on GitHub's image diff view. What changed? It's generally a good practice to provide enough detail in your commit message for people to understand the motivation. A rule of thumb is that a commit message not only explains _what_ you did, but maybe even more importantly _why_ you are doing it.

**Comment by gselzer** [comment]
> Thanks for working on this, @jackrueth! I added some additional comments beneath @gselzer's comments.
> 
> One additional thing: with [ca075cf](https://github.com/imagej/tutorials/commit/ca075cfd952b3d02d632a95d35fbfa5dd8e3127e) I don't see any significant different on the lines based on GitHub's image diff view. What changed? It's generally a good practice to provide enough detail in your commit message for people to understand the motivation. A rule of thumb is that a commit message not only explains _what_ you did, but maybe even more importantly _why_ you are doing it.

This commit changed the map (image) displayed at the bottom of the `README`. But, I agree, some more elaboration in the commit may have been helpful.

File: README.md (hunk 1 above)
Comment (gselzer) [reply README.md:21]: This is a fantastic suggestion @ctrueden. I'd add in hyperlinks/anchors to the table to make it easy to pick out the option that the user thinks best.

@jackrueth I'm happy to assist you in creating this

File: README.md (hunk 3 above)
Comment (gselzer) [reply README.md:23]: Why do we need that wiki page? It seems like it is just another thing to keep up-to-date. Why can't we just replace all references to that page with this `README`?

Comment (ctrueden) [reply README.md:23]: When you type "jupyter" into the imagej.net search bar, it is the page that comes up first. I think we should keep it, but keep it very trim, linking to the README. I have edited it accordingly.

**Comment by ctrueden** [comment]
> This commit changed the map (image) displayed at the bottom of the `README`.

Yeah, I know, but I could not tell the difference in the visual diff. I checked out the two versions of the files and compared them locally by eye, and I now see the difference I missed: the commit removes the transparency, filling it instead with white. So it does not actually changes the lines at all—only the background. With a dark theme, I see how the transparent background would be a problem.

Anyway, do we really need that directory diagram? @jackrueth Did you find it helpful? I find it redundant and not really helpful, personally. You can just, you know, look at the directory structure yourself, if you want to see it.

**Comment by gselzer** [comment]
> Anyway, do we really need that directory diagram? @jackrueth Did you find it helpful? I find it redundant and not really helpful, personally. You can just, you know, look at the directory structure yourself, if you want to see it.

This is a good question. How difficult is it for a novice user to figure out where to learn about the `gauss` Op? If you think that they could do it *easily* without the image, I think we could delete it. But I also don't know why you would delete it, outside of the extra maintenance needs added...

**Comment by ctrueden** [comment]
> I also don't know why you would delete it, outside of the extra maintenance needs added...

Exactly. Any time something is added or removed, this diagram needs to be updated. It bloats the repository.

> How difficult is it for a novice user to figure out where to learn about the gauss Op?

I don't know. But there are hundreds of ops; are we really going to have them all in a giant image chart?

What's dumb is that if you press <kbd>T</kbd> and type `gauss`, you'll find the gauss notebook. But if you type `gauss` into the search and press enter, you [get this result](https://github.com/imagej/tutorials/search?q=gauss), which does not include that notebook.

**Comment by gselzer** [comment]
> But I also don't know why you would delete it, outside of the extra maintenance needs added...

Although, the image is actually out-of-date, which makes me lean towards deleting it :laughing:

File: README.md (hunk 1 above)
Comment (jackrueth) [reply README.md:29]: I just added this table to commit 539c8fb..b2ac943 . Let me know how it looks!

File: README.md (hunk 2 above)
Comment (jackrueth) [reply README.md:38]: Added this in commit b2ac943..0c2022d . Thanks!

File: README.md (hunk 3 above)
Comment (jackrueth) [reply README.md]: Updated this link in commit 9e6b02d . Link is working well from my end.

**Comment by jackrueth** [comment]
@ctrueden @gselzer Personally, the diagram was a bit confusing as someone who had no idea how to follow the flow chart. Deleting it may be a good idea to make sure it doesn't distract first time users from more important information.

**Comment by gselzer** [comment]
> @ctrueden @gselzer Personally, the diagram was a bit confusing as someone who had no idea how to follow the flow chart. Deleting it may be a good idea to make sure it doesn't distract first time users from more important information.

Cool, so long as @elevans is okay with it, can you just remove it from the repo @jackrueth?

File: README.md (hunk 4 above)
Comment (jackrueth) [reply README.md]: Consolidated two headers from Using BeakerX and More Information on Jupyter Notebooks into one named Notebook technologies in commit 58aef7a . Let me know how it looks!

**Comment by elevans** [comment]
No problems here, remove it if it's confusing for folks.

File: README.md (hunk 1 above)
Comment (gselzer) [reply README.md:29]: @jackrueth the table looks great!

Do you think we should make the table option names consistent with the instruction title below?

Comment (gselzer) [reply README.md:29]: And, in that same vein, we should keep the ordering the same between the table and the instruction ordering below

Comment (gselzer) [reply README.md:29]: By the way @jackrueth, it looks like Github does its own markdown cell wrapping. Those `<br>` tags probably aren't necessary, but maybe they still look nice?

File: README.md (hunk 3 above)
Comment (gselzer) [reply README.md]: The link looks fine, but I'm confused as to why we are even linking there. It's just going to take us back to this `README` :laughing:

I'd argue we just remove the link.

File: README.md (hunk 4 above)
Comment (gselzer) [reply README.md]: This is very nice, thanks. I would maybe suggest using a larger header though, to distinguish this section from the earlier viewing sections.

File: README.md (hunk 1 above)
Comment (jackrueth) [reply README.md]: 1a82bbb commit fixed the titles of the methods to match their corresponding header and fc9e62c fixed the order of the headers to match the order of methods in the table.

File: README.md (hunk 3 above)
Comment (jackrueth) [reply README.md]: Haha I agree. Just realized that this is unnecessary. Fixed in commit 85c99d9.

File: README.md (hunk 4 above)
Comment (jackrueth) [reply README.md]: Agreed. Making it one size better looks better to me so it becomes more noticeable. Commit 443c57c

**Comment by jackrueth** [comment]
I removed it from the repo and all looks good to me now. Commit b75ebc6.

File: README.md (hunk 1 above)
Comment (gselzer) [reply README.md]: Thanks for cleaning this up!

One last gripe: We should have a section on viewing with nbviewer, if it is an option on the table.

I don't remember which link @ctrueden was using, but I'd go with this one:

https://nbviewer.org/github/imagej/tutorials/tree/master/

File: README.md (hunk 4 above)
Comment (gselzer) [reply README.md]: Looks great! :1st_place_medal:

File: README.md (hunk 1 above)
Comment (jackrueth) [reply README.md]: Got it. I added a header and linked it in the table. I also put in a badge that links to the nbviewer. Commit 36c10fd.

Comment (gselzer) [reply README.md]: Looks great!

**Review by gselzer — APPROVED** [review]

**Comment by jackrueth** [comment]
@ctrueden @jackrueth need to add section of "Why Jupyter" and edit intro jupyter notebook.

**Comment by jackrueth** [comment]
@ctrueden I created a header that we discussed yesterday called Why Jupyter to the README under commit 4447da7 . Wondering what you think a good description for this section would be? I added a little that I thought worked but am looking for your input.

**Comment by ctrueden** [comment]
Action items from our meeting today:

* Flesh out "Why Jupyter" blurb: "[Jupyter notebooks](https://jupyter.org/) are a great outreach tool to demonstrate how programs work interactively in real time. We use Jupyter for teaching ImageJ2 so that you can try and change the code yourself easily. See the [Why Jupyter? notebook](notebooks/0-Why-Jupyter.ipynb) if you have never used Jupyter notebook before."

* Slim down the toplevel README in favor of the notebooks and java sub-READMEs.
    * Make the toplevel README explain that both exist, and link to both.
    * Notebooks are a “guided tour” teaching ImageJ2.
    * Java code is a collection of examples and howtos for different use cases.
    * Take out the flow charts

See [the meeting Google Doc]([Jupyter notebooks](https://jupyter.org/) are a great outreach tool to demonstrate how programs work interactively in real time. We use Jupyter for teaching ImageJ2 so that you can try and change the code yourself easily. See the [Why Jupyter? notebook](notebooks/0-Why-Jupyter.ipynb) if you have never used Jupyter notebook before.) for full notes.

**Comment by ctrueden** [comment]
Whoo hoo, thanks @jackrueth!

**Pull request merged**